sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,170 @@
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey, range);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
+    // must always be less than *end*.
+    ByteKey rangeEndKey =
+        (nextKey.equals(range.getEndKey())) ? next(range.getEndKey(), range) : range.getEndKey();
+    ByteKeyRange res = ByteKeyRange.of(nextKey, rangeEndKey);
+    this.range = ByteKeyRange.of(range.getStartKey(), nextKey);
+    return res;
+  }
+
+  /**
+   * Attempts to claim the given key.
+   *
+   * <p>Must be larger than the last successfully claimed key.
+   *
+   * @return {@code true} if the key was successfully claimed, {@code false} if it is outside the
+   *     current {@link ByteKeyRange} of this tracker (in that case this operation is a no-op).
+   */
+  @Override
+  protected synchronized boolean tryClaimImpl(ByteKey key) {
+    checkArgument(
+        lastAttemptedKey == null || key.compareTo(lastAttemptedKey) > 0,
+        "Trying to claim key %s while last attempted was %s",
+        key,
+        lastAttemptedKey);
+    checkArgument(
+        key.compareTo(range.getStartKey()) > -1,
+        "Trying to claim key %s before start of the range %s",
+        key,
+        range);
+    lastAttemptedKey = key;
+    // No respective checkArgument for i < range.to() - it's ok to try claiming keys beyond
+    if (!range.getEndKey().isEmpty() && key.compareTo(range.getEndKey()) > -1) {
+      return false;
+    }
+    lastClaimedKey = key;
+    return true;
+  }
+
+  /**
+   * Marks that there are no more keys to be claimed in the range.
+   *
+   * <p>E.g., a {@link DoFn} reading a file and claiming the key of each record in the file might
+   * call this if it hits EOF - even though the last attempted claim was before the end of the
+   * range, there are no more keys to claim.
+   */
+  public synchronized void markDone() {
+    lastAttemptedKey = range.getEndKey();
+  }
+
+  @Override
+  public synchronized void checkDone() throws IllegalStateException {
+    checkState(
+        lastAttemptedKey.compareTo(range.getEndKey()) >= 0
+            || next(lastAttemptedKey, range).equals(range.getEndKey()),
+        "Last attempted key was %s in range %s, claiming work in [%s, %s) was not attempted",
+        lastAttemptedKey,
+        range,
+        next(lastAttemptedKey, range),
+        range.getEndKey());
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("range", range)
+        .add("lastClaimedKey", lastClaimedKey)
+        .add("lastAttemptedKey", lastAttemptedKey)
+        .toString();
+  }
+
+  // Utility methods
+  @VisibleForTesting
+  static ByteKey next(ByteKey key, @Nullable ByteKeyRange range) {
+    // An empty ByteKey may represent a lower or upper bound, we use the range to infer it.
+    if (key.isEmpty() && range != null) {
```

Comment:
@jkff this is also another tricky part, because ByteKey has this weirdness of [] at the beginning and at the end if its full range.